### PR TITLE
Store outdegrees in PageRank

### PR DESCRIPTION
**Pull request by abhinavmehndiratta**
Instead of calculating outdegree repeatedly,we could just store it.Just uses a bit more memory.
Some benchmarks -
```julia
julia> g = CompleteGraph(1000)
{1000, 499500} undirected simple Int64 graph

julia> @btime old_pagerank(g)
  5.114 ms (7 allocations: 23.95 KiB)

julia> @btime pagerank(g)
  2.671 ms (10 allocations: 31.94 KiB)
```

### CODE REVIEW

**Comment by SohamTamba** [comment]
Shouldn't allocations increase by `1`, not `3`?

**Comment by abhinavmehndiratta** [comment]
@SohamTamba here is another example -
```julia
julia> g = loadsnap(:facebook_combined)
{4039, 88234} undirected simple Int64 graph

julia> @benchmark pagerank(g)
BenchmarkTools.Trial: 
  memory estimate:  126.75 KiB
  allocs estimate:  14
  --------------
  minimum time:     5.996 ms (0.00% GC)
  median time:      6.097 ms (0.00% GC)
  mean time:        6.247 ms (1.62% GC)
  maximum time:     68.357 ms (90.88% GC)
  --------------
  samples:          799
  evals/sample:     1

julia> @benchmark old_pagerank(g)
BenchmarkTools.Trial: 
  memory estimate:  95.06 KiB
  allocs estimate:  10
  --------------
  minimum time:     7.356 ms (0.00% GC)
  median time:      7.469 ms (0.00% GC)
  mean time:        7.627 ms (1.57% GC)
  maximum time:     68.350 ms (88.90% GC)
  --------------
  samples:          655
  evals/sample:     1

```
Probably not a significant speedup,but for more dense graphs,the speedup is more.I have no idea why it shows more allocations than it should.

**Comment by SohamTamba** [comment]
Dense graphs have more edges so the inner loop that uses `inneighbors` has more iterations.

Another idea: How about replacing
`[outdegree(g, u) for u in vertices(g)]`
with
`[α*outdegree(g, u) for u in vertices(g)]`

You can save one multiplication in the inner loop.

**Comment by abhinavmehndiratta** [comment]
> Dense graphs have more edges so the inner loop that uses `outdegree` has more iterations.
> 
> Another idea: How about replacing
> `[outdegree(g, u) for u in vertices(g)]`
> with
> `[α*outdegree(g, u) for u in vertices(g)]`
> 
> You can save one multiplication in the inner loop per iteration.
Good point.I'll run some benchmarks.
*α/outdegree(g, u) , right ?

**Comment by SohamTamba** [comment]
Yes mistakenly wrote `α*outdegree(g, u)`

**Comment by SohamTamba** [comment]
Another idea is to compute `x_alpha_degree = [ x[u]*α/outdegree(g, u)] for u in vertices(g)` in the outer loop so the inner loop only performs addition.

This might not be worth it for a sparse graph but if consider giving it a try.
Saving work of the factor `ne(g)` but increasing work of the factor `nv(g)`

**Comment by abhinavmehndiratta** [comment]
> Another idea is to compute `x_alpha_degree = [ x[u]*α/outdegree(g, u)] for u in vertices(g)` in the outer loop so the inner loop only performs addition.
> 
> This might not be worth it for a sparse graph but if consider giving it a try.
> Saving work of the factor `ne(g)` but increasing work of the factor `nv(g)`

I thought of another,finding dangling nodes in the same loop as out_degree.

**Comment by SohamTamba** [comment]
What are `dangling_nodes` used for?
They don't seem to affect he output.

**Comment by abhinavmehndiratta** [comment]
> What are `dangling_nodes` used for?
> It doesn't seem to affect he output.
> 
> I think it can be removed.
> cc: @sbromberger

They are used to calculate the dangling_sum, used here -
https://github.com/JuliaGraphs/LightGraphs.jl/blob/072cda53f04496f27e9d6d7d8111d2157754c88e/src/centrality/pagerank.jl#L39

**Comment by abhinavmehndiratta** [comment]
But I don't get the point of dangling_weights

**Comment by SohamTamba** [comment]
`p[v]` can be replaced with 1.0/N entirely.
A waste of memory and array access (Unless I'm missing something).

**Comment by abhinavmehndiratta** [comment]
> `p[v]` can be replaced with `1.0/N` entirely.
> A waste of memory and array access (Unless I'm missing something).

True.I'll make all these changes and test.

**Comment by SohamTamba** [comment]
Nice!
Can you run benchmarks comparing 
1. Old Page Rank
2. The current PR
3. Optimise inner loop and the expense of outer loop using [x_alpha_degree](https://github.com/JuliaGraphs/LightGraphs.jl/pull/1141#issuecomment-466629523)

Run it on a big sparse graph in SNAP Datasets such as the Twitter Graph.

I think that should wrap up this PR

File: src/centrality/pagerank.jl [hunk 1]
```diff
@@ -18,31 +18,30 @@ function pagerank(
     n=100::Integer, 
     ϵ=1.0e-6
     ) where U <: Integer
-    out_degree = [outdegree(g, u) for u in vertices(g)]
-    # collect dangling nodes
-    dangling_nodes = [v for v in vertices(g) if out_degree[v] == 0]
+    α_div_outdegree = Vector{Float64}(undef,nv(g))
+    dangling_nodes = Vector{U}()
+    for v in vertices(g)
+        outdegree(g, v) == 0 && push!(dangling_nodes, v)
```
Comment (SohamTamba) [inline src/centrality/pagerank.jl]: I think LightGraphs convention does not short-circuit mutation statements like this.
cc: @sbromberger

**Comment by abhinavmehndiratta** [comment]
Here you go -
```julia
julia> g = loadsnap(:ego_twitter_d)
{81306, 1768149} directed simple Int64 graph

julia> @benchmark pagerank(g)
BenchmarkTools.Trial: 
  memory estimate:  2.11 MiB
  allocs estimate:  20
  --------------
  minimum time:     47.901 ms (0.00% GC)
  median time:      48.686 ms (0.00% GC)
  mean time:        48.964 ms (0.09% GC)
  maximum time:     58.015 ms (0.00% GC)
  --------------
  samples:          103
  evals/sample:     1

julia> @benchmark old_pagerank(g)
BenchmarkTools.Trial: 
  memory estimate:  2.11 MiB
  allocs estimate:  24
  --------------
  minimum time:     82.808 ms (0.00% GC)
  median time:      83.348 ms (0.00% GC)
  mean time:        83.671 ms (0.05% GC)
  maximum time:     87.077 ms (0.52% GC)
  --------------
  samples:          60
  evals/sample:     1
```
Using x_alpha_degree (slow cause sparse)
```julia
julia> @benchmark pagerank_mod(g)
BenchmarkTools.Trial: 
  memory estimate:  961.96 MiB
  allocs estimate:  62741981
  --------------
  minimum time:     1.471 s (2.24% GC)
  median time:      1.472 s (2.22% GC)
  mean time:        1.473 s (2.21% GC)
  maximum time:     1.479 s (2.14% GC)
  --------------
  samples:          4
  evals/sample:     1
```

**Comment by SohamTamba** [comment]
Great
The current PR is perfect.

Can you squash the 3 commits into one?
Then it can be merged

**Comment by sbromberger** [comment]
Agreed. Thanks for all the work on this and for the reviews.

**Comment by SohamTamba** [comment]
@abhinavmehndiratta would you be interested in making the same changes to `Parallel.pageerank`?

If not, I will. ^_^

**Comment by abhinavmehndiratta** [comment]
@SohamTamba I'll open a PR just a sec.

> @abhinavmehndiratta would you be interested in making the same changes to `Parallel.pageerank`?
> 
> If not, I will. ^_^

Oh forgot about that.Just a sec I'll open a PR.

**Pull request merged**